WordPress/Classes/ViewRelated/Blog/SiteManagement/SiteTagsViewController.swift

```diff
@@ -0,0 +1,376 @@
+            return
+        }
+
+        if isPerformingInitialSync {
+            setupLoadingView()
+        } else {
+            setupEmptyResultsView()
+        }
+
+        if noResultsView.superview == nil {
+            tableView.addSubview(withFadeAnimation: noResultsView)
+        }
+
+        removeSearchBar()
+    }
+
+    private func setupLoadingView() {
+        noResultsView.accessoryView = nil
+        noResultsView.titleText = loadingMessage()
+        noResultsView.messageText = ""
+        noResultsView.buttonTitle = ""
+    }
+
+    private func setupEmptyResultsView() {
+        noResultsView.accessoryView = noResultsAccessoryView()
+        noResultsView.titleText = noResultsTitle()
+        noResultsView.messageText = noResultsMessage()
+        noResultsView.buttonTitle = noResultsButtonTitle()
+        noResultsView.button.addTarget(self, action: #selector(createTag), for: .touchUpInside)
+    }
+
+    func tagsFailedLoading(error: Error) {
+        DDLogError("Tag management. Error loading tags for \(String(describing: blog.url)): \(error)")
+    }
+}
+
+// MARK: - Table view datasource
+extension SiteTagsViewController {
+    override func numberOfSections(in tableView: UITableView) -> Int {
+        return resultsController.sections?.count ?? 0
+    }
+
+    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return resultsController.sections?[section].numberOfObjects ?? 0
+    }
+
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: TableConstants.cellIdentifier, for: indexPath) as? TitleBadgeDisclosureCell, let tag = tagAtIndexPath(indexPath) else {
+            return TitleBadgeDisclosureCell()
+        }
+
+        cell.name = tag.name
+
+        if let count = tag.postCount?.intValue, count > 0 {
+            cell.count = count
+        }
+
+        return cell
+    }
+
+    fileprivate func tagAtIndexPath(_ indexPath: IndexPath) -> PostTag? {
+        return resultsController.object(at: indexPath) as? PostTag
+    }
+
+    func controllerDidChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
+        refreshNoResultsView()
+    }
+
+    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
+        return true
+    }
+
+    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
+        guard let selectedTag = tagAtIndexPath(indexPath) else {
+            return
+        }
+
+        delete(selectedTag)
+    }
+
+    private func delete(_ tag: PostTag) {
+        let tagsService = PostTagService(managedObjectContext: ContextManager.sharedInstance().mainContext)
+        refreshControl?.beginRefreshing()
+        tagsService.delete(tag, for: blog, success: {
```

Comment:
should this closure use `[weak self]`? I couldn't see a problem while running the app, but seems like it _could_ be a retain loop?